File: packages/games/pong/logic/test/vectors.test.ts
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import type { PaddleInput, PongState } from '../src/state.ts';
import { step } from '../src/step.ts';

interface Scenario {
  readonly name: string;
  readonly description: string;
  readonly inputs: string[];
  readonly checkpoints: { readonly tick: number; readonly state: PongState }[];
}

interface Vectors {
  readonly tickRate: number;
  readonly scenarios: Scenario[];
}

// Reading a fixture the sibling generator wrote. The Go conformance test
// decodes the same file into its own structs; this is the TypeScript half of
// that same trust.
// oxlint-disable-next-line typescript/no-unsafe-type-assertion
const vectors = JSON.parse(
  readFileSync(fileURLToPath(new URL('../testdata/vectors.json', import.meta.url)), 'utf8'),
) as Vectors;

function decode(symbol: string): PaddleInput {
  return { up: symbol === 'u' || symbol === 'b', down: symbol === 'd' || symbol === 'b' };
}

/**
 * Replays the recorded scenarios and checks every checkpoint.
 *
 * These same vectors are replayed by the Go implementation on the server. That
 * is the point of them: the rules exist twice, and nothing else would catch the
 * two drifting apart. A failure here means either a deliberate rule change —
 * regenerate with `pnpm --filter @littlegames/pong-logic vectors` and read the
 * diff — or an accidental one.
 */
describe('conformance vectors', () => {
  it('covers a full match played to the winning score', () => {
    const shutout = vectors.scenarios.find((scenario) => scenario.name === 'shutout');
    const final = shutout?.checkpoints.at(-1)?.state;

    expect(final?.phase).toBe('finished');
    expect(final?.winner).toBe('right');
  });

  for (const scenario of vectors.scenarios) {
    it(`replays "${scenario.name}" exactly`, () => {
      const first = scenario.checkpoints.at(0);
      if (first === undefined) {
        throw new Error(`scenario "${scenario.name}" has no checkpoints`);
      }
      expect(first.tick).toBe(0);

      let state: PongState = first.state;
      let checkpointIndex = 1;

      for (const [tick, symbols] of scenario.inputs.entries()) {
        state = step(state, {
          left: decode(symbols.slice(0, 1)),
          right: decode(symbols.slice(1, 2)),
        });

        const expected = scenario.checkpoints[checkpointIndex];
        if (expected !== undefined && expected.tick === tick + 1) {
          expect(state, `${scenario.name} diverged at tick ${String(tick + 1)}`).toEqual(
            expected.state,
          );
          checkpointIndex += 1;
        }
      }
    });
  }
});

File: server/nakama/pong/vectors_test.go
package pong

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// vectorsPath points at the file the TypeScript reference implementation
// writes. It is read, never copied: a second copy would be free to drift, which
// is the exact failure these vectors exist to prevent.
const vectorsPath = "../../../packages/games/pong/logic/testdata/vectors.json"

type checkpoint struct {
	Tick  int   `json:"tick"`
	State State `json:"state"`
}

type scenario struct {
	Name        string       `json:"name"`
	Description string       `json:"description"`
	Inputs      []string     `json:"inputs"`
	Checkpoints []checkpoint `json:"checkpoints"`
}

type vectors struct {
	TickRate  int        `json:"tickRate"`
	Scenarios []scenario `json:"scenarios"`
}

func decodeInput(symbol byte) PaddleInput {
	return PaddleInput{
		Up:   symbol == 'u' || symbol == 'b',
		Down: symbol == 'd' || symbol == 'b',
	}
}

func loadVectors(t *testing.T) vectors {
	t.Helper()

	raw, err := os.ReadFile(filepath.Clean(vectorsPath))
	if err != nil {
		t.Fatalf("read vectors: %v", err)
	}

	var loaded vectors
	if err := json.Unmarshal(raw, &loaded); err != nil {
		t.Fatalf("decode vectors: %v", err)
	}
	if len(loaded.Scenarios) == 0 {
		t.Fatal("vectors file holds no scenarios")
	}
	return loaded
}

// TestConformsToReferenceVectors replays the scenarios the TypeScript
// implementation recorded and requires identical results.
//
// The rules exist twice, once here and once in the client. Nothing else would
// catch the two drifting apart, and a drift between client prediction and
// server truth is what makes the ball appear to teleport.
func TestConformsToReferenceVectors(t *testing.T) {
	loaded := loadVectors(t)

	if loaded.TickRate != TickRate {
		t.Fatalf("tick rate mismatch: vectors %d, server %d", loaded.TickRate, TickRate)
	}

	for _, sc := range loaded.Scenarios {
		t.Run(sc.Name, func(t *testing.T) {
			if len(sc.Checkpoints) == 0 || sc.Checkpoints[0].Tick != 0 {
				t.Fatal("scenario must open with a checkpoint at tick 0")
			}

			state := sc.Checkpoints[0].State
			next := 1

			for tick, symbols := range sc.Inputs {
				if len(symbols) != 2 {
					t.Fatalf("tick %d: malformed input %q", tick, symbols)
				}

				state = Step(state, Inputs{
					Left:  decodeInput(symbols[0]),
					Right: decodeInput(symbols[1]),
				})

				if next < len(sc.Checkpoints) && sc.Checkpoints[next].Tick == tick+1 {
					if state != sc.Checkpoints[next].State {
						t.Fatalf("diverged at tick %d\n  got  %+v\n  want %+v",
							tick+1, state, sc.Checkpoints[next].State)
					}
					next++
				}
			}
		})
	}
}

// TestVectorsCoverAFullMatch guards the vectors themselves: a suite that never
// reaches the end of a match would pass while leaving scoring untested.
func TestVectorsCoverAFullMatch(t *testing.T) {
	loaded := loadVectors(t)

	for _, sc := range loaded.Scenarios {
		last := sc.Checkpoints[len(sc.Checkpoints)-1].State
		if last.Phase == PhaseFinished && last.Winner != "" {
			return
		}
	}

	t.Fatal("no scenario plays a match to its end")
}
